### PR TITLE
Added RenderItemEvent and SetupLivingBipedAnimEvent 1.16

### PR DESCRIPTION
**Pull request by MrScautHD**
+ Added RenderItemEvent (For Disable a Item Renderer in the Fhird Person)
+ Added SetupLivingBipedAnimEvent (For Player Animations)

1.18 Version: https://github.com/MinecraftForge/MinecraftForge/pull/8223

The Event RenderItemEvent allow to Cancel the Item Renderer:

The Event SetupLivingBipedAnimEvent  allow to Animate the Player or Cancel the Vanilla Animations

### CODE REVIEW

**Comment by MrScautHD** [comment]
(i see right now the code of the event call is not pushed to github), how can i fix that?

**Comment by Luis-St** [comment]
you need to run the `gradlew genPatches` task

**Comment by AterAnimAvis** [comment]
I've not reviewed or looked at this deeply but I'll also note the lack of License Headers (`gradlew licenseFormat`)

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    private int p_2291357;
+
+    public RenderItemEvent(LivingEntity entity, ItemStack heldItem, ItemCameraTransforms.TransformType transformType, MatrixStack matrixStack, IRenderTypeBuffer renderTypeBuffer, HandSide handSide, int p_2291357) {
+        this.entity = entity;
+        this.handSide = handSide;
+        this.heldItem = heldItem;
+        this.transformType = transformType;
+        this.matrixStack = matrixStack;
+        this.renderTypeBuffer = renderTypeBuffer;
+        this.p_2291357 = p_2291357;
```
Comment (Luis-St) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: which parameters is this? add a name

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.mojang.blaze3d.matrix.MatrixStack;
+import net.minecraft.client.renderer.IRenderTypeBuffer;
+import net.minecraft.client.renderer.model.ItemCameraTransforms;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.HandSide;
+import net.minecraftforge.eventbus.api.Event;
+
+public class RenderItemEvent extends Event {
```
Comment (Luis-St) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: add javadoc to the Event

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    public LivingEntity getEntity() {
+        return this.entity;
+    }
+
+    public HandSide getHandSide() {
+        return this.handSide;
+    }
+
+    public static class Post extends RenderItemEvent {
```
Comment (Luis-St) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: also missing javadoc

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        public Post(LivingEntity entity, ItemStack heldItem, ItemCameraTransforms.TransformType transformType, MatrixStack matrixStack, IRenderTypeBuffer renderTypeBuffer, HandSide handSide, int p_2291357) {
+            super(entity, heldItem, transformType, matrixStack, renderTypeBuffer, handSide, p_2291357);
+        }
+    }
+
+    public static class Pre extends RenderItemEvent {
```
Comment (Luis-St) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: also missing javadoc

**Review by Luis-St — COMMENTED** [review]
you should update this PR to 1.17.1 and then after merging port it back to 1.16

**Comment by MrScautHD** [comment]
> you should update this PR to 1.17.1 and then after merging port it back to 1.16

i port it then just to 1.17.1

**Comment by XFactHD** [comment]
> > you should update this PR to 1.17.1 and then after merging port it back to 1.16
> 
> i port it then just to 1.17.1

1.17 will not get any more PRs merged as giga told you on Discord

**Comment by MrScautHD** [comment]
> 

? what mean you exactly, so i think it does not need so many time and then can i port it.

File: patches/minecraft/net/minecraft/client/renderer/entity/layers/HeldItemLayer.java.patch [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
++import net.minecraftforge.event.RenderItemEvent;
+ 
+ @OnlyIn(Dist.CLIENT)
+ public class HeldItemLayer<T extends LivingEntity, M extends EntityModel<T> & IHasArm> extends LayerRenderer<T, M> {
+@@ -39,7 +_,7 @@
+    }
+ 
+    private void func_229135_a_(LivingEntity p_229135_1_, ItemStack p_229135_2_, ItemCameraTransforms.TransformType p_229135_3_, HandSide p_229135_4_, MatrixStack p_229135_5_, IRenderTypeBuffer p_229135_6_, int p_229135_7_) {
+-      if (!p_229135_2_.func_190926_b()) {
++      if (!p_229135_2_.func_190926_b() && MinecraftForge.EVENT_BUS.post(new RenderItemEvent(p_229135_1_, p_229135_2_, p_229135_3_, p_229135_5_, p_229135_6_, p_229135_4_, p_229135_7_))) {
```
Comment (AterAnimAvis) [inline patches/minecraft/net/minecraft/client/renderer/entity/layers/HeldItemLayer.java.patch]: A couple of things here,
- `IEventBus#post` returns true if the posted event was canceled, your logic here is inverted
- Use Fully Qualified Names `MinecraftForge` -> `net.minecraftforge.common.MinecraftForge` and remove the added imports
- This Event should really be created and fired from a method in `ForgeHooksClient` that you call here

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 4 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: (removed)

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 3 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: (removed)

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 2 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: added

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 1 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: changed

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+ * RenderItemEvent is fired if rendering a Item.
+ * <br>
+ * This event is {@link Cancelable}. <br>
+ * If the event is canceled, the item render not anymore.<br>
+ * <br>
+ * This event does not have a result. {@link HasResult}<br>
+ * <br>
+ * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+ **/
+public class RenderItemEvent extends Event {
```
Comment (AterAnimAvis) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: You haven't annotated the event `@Cancelable`.

Formatting wise, please note that forge uses brackets on new lines
```java
public class ClassName
{
    public Object method()
    {
        // ...
    }
}
```

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    private LivingEntity entity;
+    private HandSide handSide;
+    private ItemStack heldItem;
+    private ItemCameraTransforms.TransformType transformType;
+    private MatrixStack matrixStack;
+    private IRenderTypeBuffer renderTypeBuffer;
+    private int light;
```
Comment (AterAnimAvis) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: These fields should be `final`

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.event;
```
Comment (AterAnimAvis) [inline src/main/java/net/minecraftforge/event/RenderItemEvent.java]: This event should be in `net.minecraftforge.client.event`

File: patches/minecraft/net/minecraft/client/renderer/entity/layers/HeldItemLayer.java.patch (hunk 5 above)
Comment (MrScautHD) [reply patches/minecraft/net/minecraft/client/renderer/entity/layers/HeldItemLayer.java.patch]: Fixed

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 6 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: Added

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 7 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: fixed

File: src/main/java/net/minecraftforge/event/RenderItemEvent.java (hunk 8 above)
Comment (MrScautHD) [reply src/main/java/net/minecraftforge/event/RenderItemEvent.java]: changed to it

**Comment by MrScautHD** [comment]
ah small question can i also add in this Pr the SetupLivingBipedAnimEvent

(its for animation of the player here a small example):
Post:
https://user-images.githubusercontent.com/65916181/143674635-088a6921-1250-4c9a-99f3-b7e448afcce8.mp4
Pre:



or should i make a new pr for this one?

**Comment by CheckerMark79** [comment]
I don't want to disapprove this PR but I just want to point out that some of the code looks suspiciously copied from [this mod](https://github.com/MrCrayfish/Obfuscate/tree/1.16.X/src/main/java/com/mrcrayfish/obfuscate) and especially [this class](https://github.com/MrCrayfish/Obfuscate/blob/1.16.X/src/main/java/com/mrcrayfish/obfuscate/client/event/RenderItemEvent.java). It's just that the similar names and structure caught my attention. Not 100% sure though.

**Comment by MrScautHD** [comment]
> I don't want to disapprove this PR but I just want to point out that some of the code looks suspiciously copied from [this mod](https://github.com/MrCrayfish/Obfuscate/tree/1.16.X/src/main/java/com/mrcrayfish/obfuscate) and especially [this class](https://github.com/MrCrayfish/Obfuscate/blob/1.16.X/src/main/java/com/mrcrayfish/obfuscate/client/event/RenderItemEvent.java). It's just that the similar names and structure caught my attention. Not 100% sure though.

Nope i write it by my self (and i used just the names that intellij generate or the mappings), i know Obfuscate but i don't copy of it

**Comment by MrScautHD** [comment]
so im done with the PR any changed required?

**Comment by SeagullBoii** [comment]
yay no more mixins for that :D. When will this PR merge?

**Comment by MrScautHD** [comment]
> yay no more mixins for that :D. When will this PR merge?

idk, i think gigaherz has sayed after 1.18, but i hope it will merged speeder

**Comment by MrScautHD** [comment]
oh i see right now its gives a problem with  Add RenderItemEvent #7693  because he use also the same event name

**Comment by MrScautHD** [comment]
So for me is it done :), gives any requirements? if no forge can merge it :)

**Comment by sciwhiz12** [comment]
We require that PRs first be made and merged into the current latest version of Forge before a backport PR to the LTS version (1.16.x) for the same feature/bugfix is accepted, unless the PR is only applicable to the LTS version and cannot be applied to the latest version (for such reasons including the target MC/Forge system being absent in the latest version).

As this PR does not seem to only apply to the LTS version, this PR's changes must first be merged into the latest version's branch before this can be accepted into the LTS version. Additionally, due to the current circumstances with the impending release of 1.18 and the feature freeze of the `1.17.x`, such PRs to the latest version are on hold until Forge 1.18 is released, where then those PRs will target the future `1.18.x` branch as the latest version branch.

**Comment by MrScautHD** [comment]
> 



> We require that PRs first be made and merged into the current latest version of Forge before a backport PR to the LTS version (1.16.x) for the same feature/bugfix is accepted, unless the PR is only applicable to the LTS version and cannot be applied to the latest version (for such reasons including the target MC/Forge system being absent in the latest version).
> 
> As this PR does not seem to only apply to the LTS version, this PR's changes must first be merged into the latest version's branch before this can be accepted into the LTS version. Additionally, due to the current circumstances with the impending release of 1.18 and the feature freeze of the `1.17.x`, such PRs to the latest version are on hold until Forge 1.18 is released, where then those PRs will target the future `1.18.x` branch as the latest version branch.

if i understand that right i should wait on 1.18 and then make this pr on 1.18?

**Comment by sciwhiz12** [comment]
@MrScautHD: Yes. Wait on the release of Forge 1.18, then once released, please create a PR with these changes for 1.18.

**Comment by MrScautHD** [comment]
> @MrScautHD: Yes. Wait on the release of Forge 1.18, then once released, please create a PR with these changes for 1.18.

i close now this pr and make one for 1.18 ok?

**Comment by gigaherz** [comment]
No need to close. Just make a new PR for 1.18, and leave this as backport.

**Comment by MrScautHD** [comment]
> No need to close. Just make a new PR for 1.18, and leave this as backport.

ok but need i to make a new fork?

**Comment by gigaherz** [comment]
No. Just add the official repo as a remote, fetch 1.18 branch from the repo, and create a new local branch for you to work from. Basic Git, go read up how to do that.

**Comment by MrScautHD** [comment]
oh and btw does i need to make a Pr for 1.17.1?

**Comment by MrScautHD** [comment]
oh and btw i can use use 1.17 or 1.
GitHub Desktop 30 11 2021 19_23_00
16

**Comment by MrScautHD** [comment]
what ever i made now a new fork on a other account xD

**Comment by MrScautHD** [comment]
1.18 version: https://github.com/MinecraftForge/MinecraftForge/pull/8223

**Comment by MrCrayfish** [comment]
Just want to leave this here. MrScautHD did infact look and copy the events from Obfuscate (GPLv3) originally in contrary to what was mentioned in this comment (https://github.com/MinecraftForge/MinecraftForge/pull/8217#issuecomment-980544189) and made changes in https://github.com/MinecraftForge/MinecraftForge/pull/8217/commits/037abe3070ae891c5e4e86a5d10774603ed5c870 to "fix" it.  I'll leave this up to the Forge team.

**Comment by TheCurle** [comment]
Interesting.

PR closed due to violation of the Contributor's License Agreement. 

You must own or have the rights to the code you contribute to Forge.

Further PRs by MrScaut will be heavily scrutinised.

**Comment by MrCrayfish** [comment]
For context
image

**Comment by MrScautHD** [comment]
Wait wait wait first nothing of the code is anymore the same like by mrcrayfish and btw has he also sayed for him is that ok!

**Comment by MrScautHD** [comment]
> Interesting.
> 
> PR closed due to violation of the Contributor's License Agreement.
> 
> You must own or have the rights to the code you contribute to Forge.
> 
> Further PRs by MrScaut will be heavily scrutinised.

I will request a check because it false what. Mrcrayfish say the code is not anymore from him and btw it was never from him I has before just copyed the name and now is it also not more his names all is correct and not Stoallen

**Comment by MrScautHD** [comment]
> Just want to leave this here. MrScautHD did infact look and copy the events from Obfuscate (GPLv3) originally in contrary to what was mentioned in this comment ([#8217 (comment)](https://github.com/MinecraftForge/MinecraftForge/pull/8217#issuecomment-980544189)) and made changes in [037abe3](https://github.com/MinecraftForge/MinecraftForge/commit/037abe3070ae891c5e4e86a5d10774603ed5c870) to "fix" it. I'll leave this up to the Forge team.

I never copyed your code!! I just used your names and I renamed  already all!!

**Comment by MrScautHD** [comment]
THE NEW Code is 100% self made mrcrayfish speak on the old code that I copyed the 3 variable names!!

**Comment by PepperCode1** [comment]
If this is still relevant, I would recommend renaming the `RenderItemEvent` to `RenderItemInHandEvent` because items are also rendered in GUIs and as item entities but the event doesn't deal with either of these cases.

**Comment by MrScautHD** [comment]
> If this is still relevant, I would recommend renaming the `RenderItemEvent` to `RenderItemInHandEvent` because items are also rendered in GUIs and as item entities but the event doesn't deal with either of these cases.

Yea but before need forge know that I don't soolen this mod

**Comment by MrScautHD** [comment]
> Interesting.
> 
> PR closed due to violation of the Contributor's License Agreement.
> 
> You must own or have the rights to the code you contribute to Forge.
> 
> Further PRs by MrScaut will be heavily scrutinised.

Can you give me your discord then can i say you all and you see all is fine :)

**Pull request closed without merging**